### ISSUE TITLE
IMDS v2 minimum breaks existing expressions

### ISSUE THREAD

**Issue by greed42** [NixOS/amis#71]
Setting IMDS to a minimum of `v2` breaks any expression using fetchurl to read meta-data, instead you get a 401 response:

```
nix-repl> builtins.readFile (builtins.fetchurl "http://169.254.169.254/latest/meta-data")
error:
[...]
       error: unable to download 'http://169.254.169.254/latest/meta-data': HTTP error 401
```

The only way it can be turned off down-stream is via the instance configuration. There is no way (that I can find) to clear that setting when making a new image. (Nor can it be cleared on any image that I own: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ResetImageAttribute.html only allows launchPermission to be reset.) I'm using Hashicorp Packer, but a quick look at the source suggests it won't try and set it unless my configuration specifies it. So my guess is that the attribute is getting copied from the source image, 

I'm rather surprised by the AWS "set only" behaviour here; but that's what we've got to live with. So I think it would be better if we leave requiring IMDSv2 as a decision for instance configuration. (Or downstream image creators.)

**Comment by arianvp** [comment]
yeh if  you use another  AMI as a source in packer it  will probably use `ec2:CopyImage`  and will preserve the IMDSv2 requirement.

By the way is this a hypothetical issue that yo want me to address before we start using this  repo  for  publishing or something  you are running into _today_ ?  As this repo is not (yet)  publishing public AMIs. The one on the NixOS website for  23.05 shouldn't have IMDSv2 enforced yet as far as I know.

I'm  a little bit on the fence on this. IMDSv2 is the recommended thing to use as it's more  secure and the only way to support containerized workloads on  top of your  instance  (as it affects the hop limit too. imdsv1  discards packets that had  multiple hops (e.g.  through a  container interface)); and at least NixOS' amazon profile already  uses it in all its bash scripts.

Perhaps  we could register  AMIs for both? One  with  it enabled and one without?  They will use same  EBS snapshot behind  the scenes so this has no impact on cost.

**Comment by arianvp** [comment]
Ah you found them; I guess with the public search? sneaky! :D Feel free to play with them! Note that I'll start publishing them  under a different  AWS Account soon.  That account is my personal one and I'm migrating to the NixOS Foundation's one. I also  nuke all the AMIs on  my personal account  periodically so they might disappear.

I expect to complete  the migration to the  new AWS account this week.

**Comment by greed42** [comment]
Yes, I used the power of... `ls` in the cloud.

And thanks; we run on custom images even for my test rigs, so destroy as and when you need to.

**Comment by arianvp** [comment]
Heya. I consulted the AWS Docs and even when the AMI enforces IMDSv2 by default; you can still override this at the instance level when you launch the instance. Given this workaround exists I'm going to mark this as wontfix.    Amazon Linux 2023 also stopped uploading AMIs with IMDSv2 support so this feels like the right thing to do.  Amazon will also not accept images into their marketplace that don't enforce this at the AMI level.

```
aws --profile admin-nixos-village ec2 run-instances --image-id ami-00c1419be3362e995 --instance-type t3a.nano --metadata-options '{"HttpTokens": "optional"}'
```

Also note that Amazon will  start defaulting to IMDSv2 _even_ for IMDsv1 images later this year. So the only way to launch with IMDSv1 is using `--metadata-options '{"HttpTokens": "optional"}` going forward.  So you will run into the same issue you're facing either way unless you make changes to how you launch instances

https://aws.amazon.com/blogs/aws/amazon-ec2-instance-metadata-service-imdsv2-by-default/

**Comment by greed42** [comment]
Thanks for detailed answer; I'll get our instance definitions updated in advance of 24.05.
